### PR TITLE
Registration feat bis

### PR DESCRIPTION
**Pull request by anjaniacatus**

### CODE REVIEW

**Review by sheenarbw — CHANGES_REQUESTED** [review]
Good progress!

There are a few things in here that seem a little out of place. Not sure what parts of this ou re still working on.

I'm just wondering, did you consider using allauth or anything like that? It looks like there is still quite a lot of complicated code to write here.

File: core/urls.py [hunk 1]
```diff
@@ -16,9 +16,13 @@
 
 from django.contrib import admin
 from django.urls import path, include
+from django.contrib.auth import views as auth_views
+
+from custom_auth import views as custom_auth_views
 
 urlpatterns = [
     path("", include("website.urls")),
+    path("accounts/register/", custom_auth_views.register, name ="register"),
```
[the remark this answers is not in the document]
Comment (anjaniacatus) [reply core/urls.py]: True

File: tests/unit_tests/test_authentification.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        self.assertEqual(len(mail.outbox), 1)
+
+
+    def test_valid_url_to_reset_password(self):
+        u = User.objects.get(email="alice@wonderland.com")
+        response_0 = self.client.post("/accounts/password_reset/",
+                                    dict(email=u.email),
+                                   format="text/html")
+        email_lines = mail.outbox[0].body.splitlines()
+        reset_password_url = email_lines[2].replace(" ", "")
```
[the remark this answers is not in the document]
Comment (anjaniacatus) [reply tests/unit_tests/test_authentification.py:55]: also true

**Comment by anjaniacatus** [comment]
> Good progress!
> 
> There are a few things in here that seem a little out of place. Not sure what parts of this ou re still working on.
> 
> I'm just wondering, did you consider using allauth or anything like that? It looks like there is still quite a lot of complicated code to write here.

i will check that

File: templates/_base.html [hunk 3]
```diff
@@ -22,7 +23,16 @@
         style="display: none;">
 
         {% for child in link.children %}
-        <a href="{{ child.href }}" class="header_button">{{ child.label }}</a>
+            {% if child.action_url %}
+            <form method="post" action={{ child.action_url}}>
+                    {% csrf_token %}
+                    {{ form | crispy }}
```
Comment (sheenarbw) [inline templates/_base.html:29]: This is going to cause bugs.

Any time a template is rendered with with a context including a `form` then then the form will be rendered in the heading.

Eg:


image

```
from django import forms


class MyForm(forms.Form):
    name = forms.CharField(max_length=100, required=True, label="Your Name")


def page_home(request):
    context = {"form": MyForm()}
    return render(request, "page_home.html", context=context)
```

Of course this is a silly example, but we use forms in the cfp and review process. 

It looks like all you want here is a logout button that works. The logout form doesnt have any visable fields. So you could leave out the {{ form | crispy}} line entirely and then that bug will go away. 

That said, there is another downside to this approach in that we'll need to make sure this works responsively. We are going to need to make the header area responsive at some point.

There is a much simpler approach. 

Add this to the base settings:
```
ACCOUNT_LOGOUT_ON_GET = True
```

Then a normal get request to logout will log the user out. So then all the action_url functionality could be removed. 

There are downsides to letting people log out using a GET request, but I dont think we are in any danger. 

So to summarise,

I think it would be good to just use a GET to logout. But if you really want to keep the POST then it's important to remove the `{{form | crispy}}` line

File: tests/unit_tests/test_authentification.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
+import pytest
+from django.core import mail
+from django.test import Client
+from pytest_django.asserts import assertRedirects, assertTemplateUsed
+from custom_auth.models import User
+
+client = Client()
+
+@pytest.fixture
+def user_fixture(db, django_user_model):
```
[the remark this answers is not in the document]
Comment (sheenarbw) [reply tests/unit_tests/test_authentification.py:10]: Also, it would be good to put test fixtures in conftest.py in case we want to use them again in different test files

File: tests/unit_tests/test_authentification.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+@pytest.mark.django_db
+def test_register():
+    response = client.post(
+        "/accounts/register/",
+        dict(email="rabbit.white@wonderland.com",
+        last_name="Rabbit", first_name="White",
+        password1="FNEYghfr",
+        password2="FNEYghfr"),
+        format="text/html")
+    assert response.status_code == 404
```
Comment (sheenarbw) [inline tests/unit_tests/test_authentification.py:28]: I'm not sure why you are testing that this page does not exist... There are a lot of urls that would give us a 404

**Review by sheenarbw — CHANGES_REQUESTED** [review]
Sorry this took so long to review. Life has been very busy!

I noticed a few weird things around database setup. You tweaked some settings in a way that breaks the way the dev_db was being used.  I added some info about it.

And there are some layout UI things. And something that will definately turn into a bug later. 

Mostly it looks solid, it just needs a rouch of polish then we're good to go!

File: .github/workflows/coverage.yml [hunk 6]
```diff
@@ -17,11 +17,11 @@ jobs:
 
     services:
       postgres:
-        image: postgres:10.8
+        image: postgres:14.13
```
[the remark this answers is not in the document]
Comment (anjaniacatus) [reply .github/workflows/coverage.yml:20]: ok i didn't notice that. I should have asked effectively.

[the remark this answers is not in the document]
Comment (anjaniacatus) [reply .github/workflows/coverage.yml:20]: @sheenarbw  on the branch main we have 14.13 . so what is the right version then?
https://github.com/djangocon/2025.djangocon.africa/blob/main/.github/workflows/django.yml

File: dev_db/docker-entrypoint-initdb.d/create-test-db.sql [hunk 7]
```diff
@@ -1,3 +1,3 @@
-CREATE DATABASE test_db;
-grant all privileges on database test_db to pguser;
+CREATE DATABASE db;
```
[the remark this answers is not in the document]
Comment (anjaniacatus) [reply dev_db/docker-entrypoint-initdb.d/create-test-db.sql]: I have  put it back

File: templates/_base.html (hunk 3 above)
Comment (anjaniacatus) [reply templates/_base.html:29]: fixed

File: tests/unit_tests/test_authentification.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (anjaniacatus) [reply tests/unit_tests/test_authentification.py:10]: ah. thank you for the clarification

File: website/header_menu_items.py [hunk 8]
```diff
@@ -9,7 +9,7 @@ class HeaderLink:
     label: str
     href: str = ""
     children: List["HeaderLink"] = field(default_factory=list)
-
+    action_url: str = ""
```
Comment (sheenarbw) [inline website/header_menu_items.py:12]: If this isn't being used it would be good to remove it.

File: templates/_base.html (hunk 3 above)
Comment (sheenarbw) [reply templates/_base.html:29]: This isn't fixed. It still has `{{ form | crispy }}` in the header area. So if any view has a contect that includes something named `form` then it will show up in the header area. 

This will be bad because we use a lot of forms. So a lot of pages will end up with a bug.

Please remove the form from the header area. 

Take a look at the comment I left before. If you read from the top then you'll see a demonstration of the bug.

Comment (sheenarbw) [reply templates/_base.html:29]: If the action_url stuff isn't being used then it would be good to remove it. Then it wont be able to cause any bugs.

**Review by sheenarbw — CHANGES_REQUESTED** [review]
Looking good!

There is still a bug in the action_url functionality, but since that isn't being used it can just be removed.

File: templates/_base.html (hunk 3 above)
Comment (anjaniacatus) [reply templates/_base.html]: i've forgot to clean it. now it's ok

**Review by sheenarbw — APPROVED** [review]

**Comment by anjaniacatus** [comment]
@sheenarbw i've fixed all of formatting issue
